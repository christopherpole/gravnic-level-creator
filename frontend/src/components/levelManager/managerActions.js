import React from 'react';
import styled from 'styled-components';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types';

import {
  createNewLevel,
  loadLevel,
  saveLevel,
  deleteSelectedLevel,
  copyLevel,
  beginRenameLevel,
  finishRenameLevel,
} from '../../actions/levelManager';
import Button from '../common/button';
import { getLevelManagerButtonDisabledStates } from '../../selectors';

export const Wrapper = styled.div`
  border-top: 1px solid white;
  padding: calc(${props => props.theme.structureSpacing} / 2);
`;

export const Toolbar = styled.ul`
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-gap: calc(${props => props.theme.structureSpacing} / 2);
`;

export const ToolbarAction = styled.li`
  list-style-type: none;
  flex-grow: 1;
`;

export const StyledButton = styled(Button)`
  width: 100%;
`;

export const ManagerActions = ({
  renamingLevelId,
  createLevelAction,
  loadLevelAction,
  saveLevelAction,
  deleteSelectedLevelAction,
  copyLevelAction,
  beginRenameLevelAction,
  finishRenameLevelAction,
  buttonDisabledStates,
}) => (
  <Wrapper>
    <Toolbar>
      <ToolbarAction>
        <StyledButton
          disabled={buttonDisabledStates.btnNew}
          onClick={createLevelAction}
          id="btn-new"
        >
          <span>New</span>
        </StyledButton>
      </ToolbarAction>
      <ToolbarAction>
        <StyledButton
          disabled={buttonDisabledStates.btnLoad}
          onClick={loadLevelAction}
          id="btn-load"
        >
          <span>Load</span>
        </StyledButton>
      </ToolbarAction>
      <ToolbarAction>
        <StyledButton
          disabled={buttonDisabledStates.btnSave}
          onClick={saveLevelAction}
          id="btn-save"
        >
          <span>Save</span>
        </StyledButton>
      </ToolbarAction>
      <ToolbarAction>
        <StyledButton
          disabled={buttonDisabledStates.btnDelete}
          onClick={deleteSelectedLevelAction}
          id="btn-delete"
        >
          <span>Delete</span>
        </StyledButton>
      </ToolbarAction>
      <ToolbarAction>
        <StyledButton
          disabled={buttonDisabledStates.btnCopy}
          onClick={copyLevelAction}
          id="btn-copy"
        >
          <span>Copy</span>
        </StyledButton>
      </ToolbarAction>
      <ToolbarAction>
        {!renamingLevelId && (
          <StyledButton
            disabled={buttonDisabledStates.btnRename}
            onClick={beginRenameLevelAction}
            id="btn-rename"
          >
            <span>Rename</span>
          </StyledButton>
        )}
        {renamingLevelId && (
          <StyledButton onClick={finishRenameLevelAction} id="btn-done">
            <span>Done</span>
          </StyledButton>
        )}
      </ToolbarAction>
    </Toolbar>
  </Wrapper>
);

ManagerActions.defaultProps = {
  renamingLevelId: null,
};

ManagerActions.propTypes = {
  createLevelAction: PropTypes.func.isRequired,
  loadLevelAction: PropTypes.func.isRequired,
  saveLevelAction: PropTypes.func.isRequired,
  deleteSelectedLevelAction: PropTypes.func.isRequired,
  copyLevelAction: PropTypes.func.isRequired,
  beginRenameLevelAction: PropTypes.func.isRequired,
  finishRenameLevelAction: PropTypes.func.isRequired,
  renamingLevelId: PropTypes.string,
  buttonDisabledStates: PropTypes.shape({
    btnNew: PropTypes.bool.isRequired,
    btnLoad: PropTypes.bool.isRequired,
    btnSave: PropTypes.bool.isRequired,
    btnDelete: PropTypes.bool.isRequired,
    btnCopy: PropTypes.bool.isRequired,
    btnRename: PropTypes.bool.isRequired,
  }).isRequired,
};

const mapStateToProps = state => ({
  renamingLevelId: state.levelManager.renamingLevelId,
  buttonDisabledStates: getLevelManagerButtonDisabledStates(state),
});

const mapDispatchToProps = dispatch => ({
  createLevelAction: bindActionCreators(createNewLevel, dispatch),
  loadLevelAction: bindActionCreators(loadLevel, dispatch),
  saveLevelAction: bindActionCreators(saveLevel, dispatch),
  deleteSelectedLevelAction: bindActionCreators(deleteSelectedLevel, dispatch),
  copyLevelAction: bindActionCreators(copyLevel, dispatch),
  beginRenameLevelAction: bindActionCreators(beginRenameLevel, dispatch),
  finishRenameLevelAction: bindActionCreators(finishRenameLevel, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(ManagerActions);
