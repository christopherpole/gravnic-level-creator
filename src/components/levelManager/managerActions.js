import React from 'react';
import styled from 'styled-components';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import PropTypes from 'prop-types';

import {
  createLevel,
  loadLevel,
  saveLevel,
  deleteLevel,
  copyLevel,
  beginRenameLevel,
  finishRenameLevel,
} from '../../actions/levelManager';

export const Wrapper = styled.div`
  border-top: 1px solid white;
  padding: calc(${props => props.theme.structureSpacing} / 2);
`;

export const Toolbar = styled.ul`
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-gap: calc(${props => props.theme.structureSpacing} / 2);
`;

export const ToolbarAction = styled.li`
  list-style-type: none;
  flex-grow: 1;
`;

export const Button = styled.button`
  padding: calc(${props => props.theme.structureSpacing} / 4)
    calc(${props => props.theme.structureSpacing} / 2);
  width: 100%;
  cursor: pointer;
`;

export const ManagerActions = ({
  selectedLevelId,
  renamingLevelId,
  createLevelAction,
  loadLevelAction,
  saveLevelAction,
  deleteLevelAction,
  copyLevelAction,
  beginRenameLevelAction,
  finishRenameLevelAction,
}) => (
  <Wrapper>
    <Toolbar>
      <ToolbarAction>
        <Button disabled={renamingLevelId} onClick={createLevelAction}>
          <span>New</span>
        </Button>
      </ToolbarAction>
      <ToolbarAction>
        <Button disabled={!selectedLevelId || renamingLevelId} onClick={loadLevelAction}>
          <span>Load</span>
        </Button>
      </ToolbarAction>
      <ToolbarAction>
        <Button disabled={!selectedLevelId || renamingLevelId} onClick={saveLevelAction}>
          <span>Save</span>
        </Button>
      </ToolbarAction>
      <ToolbarAction>
        <Button disabled={!selectedLevelId || renamingLevelId} onClick={deleteLevelAction}>
          <span>Delete</span>
        </Button>
      </ToolbarAction>
      <ToolbarAction>
        <Button disabled={!selectedLevelId || renamingLevelId} onClick={copyLevelAction}>
          <span>Copy</span>
        </Button>
      </ToolbarAction>
      <ToolbarAction>
        {!renamingLevelId && (
          <Button disabled={!selectedLevelId} onClick={beginRenameLevelAction}>
            <span>Rename</span>
          </Button>
        )}
        {renamingLevelId && (
          <Button onClick={finishRenameLevelAction}>
            <span>Done</span>
          </Button>
        )}
      </ToolbarAction>
    </Toolbar>
  </Wrapper>
);

ManagerActions.defaultProps = {
  selectedLevelId: null,
  renamingLevelId: null,
};

ManagerActions.propTypes = {
  createLevelAction: PropTypes.func.isRequired,
  loadLevelAction: PropTypes.func.isRequired,
  saveLevelAction: PropTypes.func.isRequired,
  deleteLevelAction: PropTypes.func.isRequired,
  copyLevelAction: PropTypes.func.isRequired,
  beginRenameLevelAction: PropTypes.func.isRequired,
  finishRenameLevelAction: PropTypes.func.isRequired,
  selectedLevelId: PropTypes.string,
  renamingLevelId: PropTypes.string,
};

const mapStateToProps = state => ({
  selectedLevelId: state.levelManager.selectedLevelId,
  renamingLevelId: state.levelManager.renamingLevelId,
});

const mapDispatchToProps = dispatch => ({
  createLevelAction: bindActionCreators(createLevel, dispatch),
  loadLevelAction: bindActionCreators(loadLevel, dispatch),
  saveLevelAction: bindActionCreators(saveLevel, dispatch),
  deleteLevelAction: bindActionCreators(deleteLevel, dispatch),
  copyLevelAction: bindActionCreators(copyLevel, dispatch),
  beginRenameLevelAction: bindActionCreators(beginRenameLevel, dispatch),
  finishRenameLevelAction: bindActionCreators(finishRenameLevel, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(ManagerActions);
