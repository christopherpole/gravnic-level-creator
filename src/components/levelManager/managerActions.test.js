import React from 'react';
import { configure, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';
import { spy } from 'sinon';

import { ManagerActions, Button } from './managerActions';

configure({ adapter: new Adapter() });

describe('The level manager actions', () => {
  let props;

  beforeEach(() => {
    props = {
      loadLevelAction: () => {},
      saveLevelAction: () => {},
    };
  });

  it('Renders without exploding', () => {
    const managerActions = shallow(<ManagerActions {...props} />);

    expect(managerActions).toHaveLength(1);
  });

  it('Matches the current snapshot', () => {
    const managerActions = shallow(<ManagerActions {...props} />);

    expect(toJson(managerActions)).toMatchSnapshot();
  });

  it('Fires the load level action when clicking on a level', () => {
    const loadLevelSpy = spy();
    const managerActions = shallow(<ManagerActions {...props} loadLevelAction={loadLevelSpy} />);
    const button = managerActions.find(Button).at(0);
    button.simulate('click');

    expect(loadLevelSpy.calledOnce).toBe(true);
  });

  it('Fires the save level action when clicking on a level', () => {
    const saveLevelSpy = spy();
    const managerActions = shallow(<ManagerActions {...props} saveLevelAction={saveLevelSpy} />);
    const button = managerActions.find(Button).at(1);
    button.simulate('click');

    expect(saveLevelSpy.calledOnce).toBe(true);
  });
});
