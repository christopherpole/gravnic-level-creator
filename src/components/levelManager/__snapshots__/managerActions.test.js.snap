// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The level manager actions Matches the current snapshot 1`] = `
<styled.div>
  <styled.ul>
    <styled.li>
      <styled.button
        disabled={false}
        onClick={[Function]}
      >
        Load
      </styled.button>
    </styled.li>
    <styled.li>
      <styled.button
        disabled={false}
        onClick={[Function]}
      >
        Save
      </styled.button>
    </styled.li>
    <styled.li>
      <styled.button
        disabled={false}
        onClick={[Function]}
      >
        Delete
      </styled.button>
    </styled.li>
    <styled.li>
      <styled.button
        disabled={false}
        onClick={[Function]}
      >
        Copy
      </styled.button>
    </styled.li>
  </styled.ul>
</styled.div>
`;

exports[`The level manager actions Matches the current snapshot if a level isn't selected 1`] = `
<styled.div>
  <styled.ul>
    <styled.li>
      <styled.button
        disabled={true}
        onClick={[Function]}
      >
        Load
      </styled.button>
    </styled.li>
    <styled.li>
      <styled.button
        disabled={true}
        onClick={[Function]}
      >
        Save
      </styled.button>
    </styled.li>
    <styled.li>
      <styled.button
        disabled={true}
        onClick={[Function]}
      >
        Delete
      </styled.button>
    </styled.li>
    <styled.li>
      <styled.button
        disabled={true}
        onClick={[Function]}
      >
        Copy
      </styled.button>
    </styled.li>
  </styled.ul>
</styled.div>
`;
